echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/PipelineCache.java

```diff
@@ -142,25 +143,68 @@ void pollPipelineConfigs() {
     }
   }
 
-  private List<Pipeline> fetchHydratedPipelines() {
-    List<Map<String, Object>> rawPipelines = front50.getPipelines();
-    if (rawPipelines == null) {
-      return Collections.emptyList();
-    }
-
+  private Map<String, Object> hydrate(Map<String, Object> rawPipeline) {
     Predicate<Map<String, Object>> isV2Pipeline = p -> {
       return p.getOrDefault("type", "").equals("templatedPipeline") &&
         p.getOrDefault("schema", "").equals("v2");
     };
 
-    return rawPipelines.stream()
-      .map((Map<String, Object> p) -> planPipelineIfNeeded(p, isV2Pipeline))
+    return planPipelineIfNeeded(rawPipeline, isV2Pipeline);
+  }
+
+  // converts a raw pipeline config from front50 into a processed Pipeline object
+  private Optional<Pipeline> process(Map<String, Object> rawPipeline) {
+    return Stream.of(rawPipeline)
+      .map(this::hydrate)
       .filter(m -> !m.isEmpty())
-      .map(m -> convertToPipeline(m))
+      .map(this::convertToPipeline)
       .filter(Objects::nonNull)
+      .map(PipelineCache::decorateTriggers)
+      .findFirst();
+  }
+
+  private List<Map<String, Object>> fetchRawPipelines() {
+    List<Map<String, Object>> rawPipelines = front50.getPipelines();
+    return (rawPipelines == null) ? Collections.emptyList() : rawPipelines;
+  }
+
+  private List<Pipeline> fetchHydratedPipelines() {
+    List<Map<String, Object>> rawPipelines = fetchRawPipelines();
+
+    return rawPipelines.stream()
+      .map(this::process)
+      .filter(Optional::isPresent)
+      .map(Optional::get)
       .collect(Collectors.toList());
   }
 
+  // looks up the latest version in front50 of a (potentially stale) pipeline config from the cache
+  // if anything fails during the refresh, we fall back to returning the cached version
+  public Pipeline refresh(Pipeline cached) {
+    try {
+      List<Map<String, Object>> latestVersion = front50.getLatestVersion(cached.getId());
```

Comment:
Think it is worth a timer or counter around this call? I'm curious about how frequently we will end up calling this... if it turns out to be a lot we might need to revisit this with something that loads in bulk for all the matchingPipelines (but maybe hopefully not because this way is definitely clearer and simpler)

Reply:
I was thinking about that. Ideally we would be using an instrumented okhttp client that already publishes RPC metrics, that would avoid a proliferation of handmade counters. 

EDIT: yes we do, the timer is `okhttp.requests` and it is tagged by `requestHost`, `statusCode`, etc.